### PR TITLE
Respond with Head to all Get requests

### PR DESCRIPTION
**Pull request by codedmart**
Still a WIP.

### CODE REVIEW

**Comment by codedmart** [comment]
@alpmestan I have basic working code locally, I will ping you tomorrow with a couple thoughts/questions.

**Comment by codedmart** [comment]
@alpmestan @jkarni any suggestions? I didn't pull out any of the commonalities from those router methods.

File: servant-server/src/Servant/Server/Internal.hs [hunk 1]
```diff
@@ -129,6 +129,13 @@ methodRouter :: (AllCTRender ctypes a)
 methodRouter method proxy status action = LeafRouter route'
   where
     route' request respond
+      | pathIsEmpty request && method == methodGet && requestMethod request == methodHead = do
+          runAction action respond $ \ output -> do
+            let accH = fromMaybe ct_wildcard $ lookup hAccept $ requestHeaders request
+            case handleAcceptH proxy (AcceptHeader accH) output of
+              Nothing -> failWith UnsupportedMediaType
+              Just (contentT, _) -> succeedWith $
+                responseLBS status [ ("Content-Type" , cs contentT)] ""
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: Can we maybe abstract this code a little, given that it's almost exactly the same as the stuff just below?

**Comment by jkarni** [comment]
Great idea! Abstracting a little the commonalities would be nice, though...

**Comment by codedmart** [comment]
@jkarni I agree, I will look at that. Thanks!

**Comment by codedmart** [comment]
@jkarni Not sure if this feels better but let me know.

File: servant-server/src/Servant/Server/Internal.hs [hunk 2]
```diff
@@ -121,6 +123,30 @@ instance (KnownSymbol capture, FromText a, HasServer sublayout)
                Just v   -> feedTo subserver v)
     where captureProxy = Proxy :: Proxy (Capture capture a)
 
+allowedMethodHead :: Method -> Request -> Bool
+allowedMethodHead method request = method == methodGet && requestMethod request == methodHead
+
+allowedMethod :: Method -> Request -> Bool
+allowedMethod method request
+  | allowedMethodHead method request = True
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: can't this whole function be:

```
allowedMethod method request = allowedMethodHead method request || requestMethod request == method
```

File: servant-server/src/Servant/Server/Internal.hs [hunk 3]
```diff
@@ -121,6 +123,30 @@ instance (KnownSymbol capture, FromText a, HasServer sublayout)
                Just v   -> feedTo subserver v)
     where captureProxy = Proxy :: Proxy (Capture capture a)
 
+allowedMethodHead :: Method -> Request -> Bool
+allowedMethodHead method request = method == methodGet && requestMethod request == methodHead
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: I'd make this a `where/let` clause in `allowedMethod`

File: servant-server/src/Servant/Server/Internal.hs [hunk 4]
```diff
@@ -121,6 +123,30 @@ instance (KnownSymbol capture, FromText a, HasServer sublayout)
...
+                    -> Request -> RouteResult Response
+processMethodRouter handleA status method headers request = case handleA of
+  Nothing -> failWith UnsupportedMediaType
+  Just (contentT, body) -> succeedWith $
+    responseLBS status hdrs bdy
+    where
+      bdy = case allowedMethodHead method request of
+        True -> ""
+        False -> body
+      hdrs = case headers of
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: `hdrs = (hContentType, cs contentT):(maybeToList headers)`

File: servant-server/src/Servant/Server/Internal.hs [hunk 5]
```diff
@@ -644,9 +667,9 @@ instance (KnownSymbol sym, HasServer sublayout)
                     Just Nothing  -> True  -- param is there, with no value
                     Just (Just v) -> examine v -- param with a value
                     Nothing       -> False -- param not in the query string
-  
+
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: I wouldn't add a blank line here.

File: servant-server/src/Servant/Server/Internal.hs [hunk 6]
```diff
@@ -147,17 +171,16 @@ methodRouterHeaders :: (GetHeaders (Headers h v), AllCTRender ctypes v)
...
-          case handleAcceptH proxy (AcceptHeader accH) (getResponse output) of
-            Nothing -> failWith UnsupportedMediaType
-            Just (contentT, body) -> succeedWith $
-              responseLBS status ( ("Content-Type" , cs contentT) : headers) body
+              handleA = handleAcceptH proxy (AcceptHeader accH) (getResponse output)
+          processMethodRouter handleA status method (Just headers) request
       | pathIsEmpty request && requestMethod request /= method =
           respond $ failWith WrongMethod
       | otherwise = respond $ failWith NotFound
+      where
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: Is this an extra `where`?

File: servant-server/src/Servant/Server/Internal.hs [hunk 7]
```diff
@@ -121,6 +123,30 @@ instance (KnownSymbol capture, FromText a, HasServer sublayout)
...
+  | requestMethod request == method = True
+  | otherwise = False
+
+processMethodRouter :: forall a. ConvertibleStrings a B.ByteString
+                    => Maybe (a, BL.ByteString) -> Status -> Method
+                    -> Maybe [(HeaderName, B.ByteString)]
+                    -> Request -> RouteResult Response
+processMethodRouter handleA status method headers request = case handleA of
+  Nothing -> failWith UnsupportedMediaType
+  Just (contentT, body) -> succeedWith $
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Internal.hs]: `responseLBS ...` on the same line as `succeedWith`

**Comment by jkarni** [comment]
Minor nitpicks, but looks good.

File: servant-server/src/Servant/Server/Internal.hs (hunk 5 above)
Comment (codedmart) [inline servant-server/src/Servant/Server/Internal.hs]: Didn't add a blank just removed white space. But. I am remove the whole line.

File: servant-server/src/Servant/Server/Internal.hs (hunk 6 above)
Comment (codedmart) [inline servant-server/src/Servant/Server/Internal.hs]: Yup oops.

**Comment by codedmart** [comment]
@jkarni thanks. I will get these changes made.

**Comment by codedmart** [comment]
@jkarni again thanks for the suggestions. Let me know if I missed anything.

**Comment by jkarni** [comment]
Awesome, thanks!

**Pull request merged**